src/listeners.js

```diff
@@ -0,0 +1,180 @@
+const assert = require('assert');
+const debug = require('debug')('listeners');
+const pulse = require('taskcluster-lib-pulse');
+const _ = require('lodash');
+const {serialize} = require('async-decorators');
+
+/**
+ * Create pulse client and consumers to trigger hooks with pulse messages 
+ *
+ * options:
+ * {
+ *   Hook:               // Azure tables for hooks
+ *   taskcreator:        // A TaskCreator instance
+ *   client:            // A tc-lib-pulse client instance
+ * }
+ */
+
+class HookListeners {
+  constructor(options) {
+    assert(options.client, 'tc-lib-pulse client must be provided');
+
+    this.taskcreator = options.taskcreator;
+    this.Hook = options.Hook;
+    this.Queues = options.Queues;
+    this.client = options.client;
+    this.pulseHookChangedListener = null;
+    this.listeners = null;
+  }
+
+  /**
+   * Setup a new pulse client using the credentials
+   * Additionally create pulse consumers for the exchanges - 
+   * `hook-created, `hook-updated` and  `hook-deleted`
+  */
+  async setup() {
+    debug('Setting up the listeners');
+    assert(this.listeners === null, 'Cannot setup twice');
+
+    const client = this.client;
+    let consumer = await pulse.consume({
+      client,      
+      bindings: [{
+        exchange: 'exchange/taskcluster-hooks/v1/hook-created',
+        routingKeyPattern: '#',
+      }, {
+        exchange: 'exchange/taskcluster-hooks/v1/hook-updated',
+        routingKeyPattern: '#',
+      }, {
+        exchange: 'exchange/taskcluster-hooks/v1/hook-deleted',
+        routingKeyPattern: '#',
+      }],
+      queueName: 'hookChanged',
+      maxLength : 50,
+    }, (msg) => this.reconcileConsumers()
+    );
+    debug('Listening to hook exchanges');
+    this.pulseHookChangedListener = consumer;
+    this.listeners = [];
+  }
+
+  /** Create a new pulse consumer for a hook */
+  async createListener({hook, oldBindings}) {
+    this.hook = hook;
+    const client =  this.client;
+    const queueName = `${hook.hookGroupId}/${hook.hookId}`; // serves as unique id for every listener
+    const listener = await pulse.consume({
+      client,      
+      queueName: queueName, 
+      maxLength : 50,
+    }, async ({payload}) => {
+      const hook = this.hook;
+      // Pass empty functions in res which are executed by `triggerHookCommon`
+      // This is more of a hack so that if the payload doesnt validate against
+      // `triggerSchema`,git  we can simply ignore the message.
+      await this.taskcreator.fire(hook, {firedBy:'pulseMessage', payload});
+    });
+    if (!this.client.isFakeClient) {
+      for (let {exchange, routingKeyPattern} of oldBindings) {
+        await this.client.withChannel(async channel => channel.unbindQueue(queueName, exchange, routingKeyPattern));
+      }
+      for (let {exchange, routingKeyPattern} of hook.bindings) {
+        await this.client.withChannel(async channel => {channel.bindQueue(queueName, exchange, routingKeyPattern);});
+      }
```

Comment:
For the case where the user adds a binding, this will in fact unbind all bindings, then re-bind them (along with the new one).  Some clever use of the `Set` class could probably avoid this, so that the only change to the queue in such a case is to call `bindQueue` for the new binding.

Reply:
I found  `_.intersectionWith(oldBindings, newBindings, _.isEqual)` and used its diff with old and new bindings to unbind and bind respectively.